org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.java

```diff
@@ -0,0 +1,1196 @@
+			return Collections.emptyList();
+		}
+		return workspaceManager.doRead(uri,
+				(doc, res) -> documentSymbolService.getDefinitions(doc, res, params, resourceAccess, cancelIndicator));
+	}
+
+	@Override
+	public CompletableFuture<List<? extends Location>> references(ReferenceParams params) {
+		return requestManager.runRead(cancelIndicator -> references(params, cancelIndicator));
+	}
+
+	/**
+	 * Compute the references. Executed in read request.
+	 */
+	protected List<? extends Location> references(ReferenceParams params, CancelIndicator cancelIndicator) {
+		URI uri = getURI(params);
+		DocumentSymbolService documentSymbolService = getService(uri, DocumentSymbolService.class);
+		if (documentSymbolService == null) {
+			return Collections.emptyList();
+		}
+		return workspaceManager.doRead(uri, (document, resource) -> documentSymbolService.getReferences(document,
+				resource, params, resourceAccess, workspaceManager.getIndex(), cancelIndicator));
+	}
+
+	@Override
+	public CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> documentSymbol(
+			DocumentSymbolParams params) {
+		return requestManager.runRead((cancelIndicator) -> documentSymbol(params, cancelIndicator));
+	}
+
+	/**
+	 * Compute the symbol information. Executed in a read request.
+	 */
+	protected List<Either<SymbolInformation, DocumentSymbol>> documentSymbol(DocumentSymbolParams params,
+			CancelIndicator cancelIndicator) {
+		URI uri = getURI(params.getTextDocument());
+		IDocumentSymbolService documentSymbolService = getIDocumentSymbolService(getResourceServiceProvider(uri));
+		if (documentSymbolService == null) {
+			return Collections.emptyList();
+		}
+		return workspaceManager.doRead(uri,
+				(document, resource) -> documentSymbolService.getSymbols(document, resource, params, cancelIndicator));
+	}
+
+	/**
+	 * @since 2.16
+	 */
+	protected IDocumentSymbolService getIDocumentSymbolService(IResourceServiceProvider serviceProvider) {
+		if (serviceProvider == null) {
+			return null;
+		}
+		if (isHierarchicalDocumentSymbolSupport()) {
+			return serviceProvider.get(HierarchicalDocumentSymbolService.class);
+		} else {
+			return serviceProvider.get(DocumentSymbolService.class);
+		}
+	}
+
+	/**
+	 * {@code true} if the {@code TextDocumentClientCapabilities} explicitly declares the hierarchical document symbol
+	 * support at LS initialization time. Otherwise, false.
+	 */
+	protected boolean isHierarchicalDocumentSymbolSupport() {
+		ClientCapabilities capabilities = initializeParams.getCapabilities();
+		if (capabilities != null) {
+			TextDocumentClientCapabilities textDocument = capabilities.getTextDocument();
+			if (textDocument != null) {
+				DocumentSymbolCapabilities documentSymbol = textDocument.getDocumentSymbol();
+				if (documentSymbol != null) {
+					Boolean hierarchicalDocumentSymbolSupport = documentSymbol.getHierarchicalDocumentSymbolSupport();
+					if (hierarchicalDocumentSymbolSupport != null) {
```

Comment:
I'd prefer checking for null and return false instead of the if-cascade

Reply:
Done